Add Xiaomi_miio controls for humidifier - led_brightness

jbouwh opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Main aim here is to add missing sensors and controls for Xiaomi humidifiers:

This PR covers:
- led brightness control for Humidifiers using the select platform

The implementation for new sensors, switched `actual_speed`, `buzzer`, `child_lock`, `dry_mode` and `clean_mode` and `motor_speed`, is handled in other PR's.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

jbouwh committed:
Add xiaomi_miio humidifier entity and sensors (#51884)

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>
Co-authored-by: Maciej Bieniek <bieniu@users.noreply.github.com>
Co-authored-by: Teemu R. <tpr@iki.fi>

jbouwh committed:
Merge branch 'rework-xiaomi-miio-fans-humidifiers' of https://github.com/home-assistant/core into rework-xiaomi-miio-fans-humidifiers

jbouwh committed:
Moved rework to select platform PR

jbouwh committed:
Add select.py to .coveragerc

jbouwh committed:
revert removal remote.py from .coveragerc

bieniu commented on homeassistant/components/xiaomi_miio/select.py:
```diff
@@ -0,0 +1,253 @@
+                    )
+                )
+
+    async_add_entities(entities, update_before_add=False)
```
Default value of `update_before_add` is `False`.
```suggestion
    async_add_entities(entities)
```

bieniu commented on homeassistant/components/xiaomi_miio/select.py:
```diff
@@ -0,0 +1,253 @@
+    def __init__(self, name, device, entry, unique_id, selector, coordinator):
+        """Initialize the generic Xiaomi attribute selector."""
+        super().__init__(name, device, entry, unique_id, coordinator)
+        self._state = None
```
`self._state` is not used.
```suggestion
```

bieniu commented on homeassistant/components/xiaomi_miio/select.py:
```diff
@@ -0,0 +1,253 @@
+    @property
+    def current_option(self):
+        """Return the current option."""
+        if not self.available and self.coordinator.data:
+            self._available = True
+            self._current_option = self._extract_value_from_attribute(
+                self.coordinator.data, self._controller.short_name
+            )
+        return (
+            getattr(self, SERVICE_TO_METHOD[self._controller.service]["property"])
+            if self.available
+            else None
+        )
```
This is not necessary.
```suggestion
        return getattr(self, SERVICE_TO_METHOD[self._controller.service]["property"])
```

jbouwh replied:
Do we not need to check for self.available here?

bieniu commented on homeassistant/components/xiaomi_miio/select.py:
```diff
@@ -0,0 +1,253 @@
+    @callback
+    def _handle_coordinator_update(self):
+        """Fetch state from the device."""
+        # On state change the device doesn't provide the new state immediately.
+        state = self.coordinator.data
+        if not state:
+            return
+        _LOGGER.debug("Got new state: %s", state)
+        self._available = True
+        self._current_option = self._extract_value_from_attribute(
+            state, self._controller.short_name
+        )
+        self.async_write_ha_state()
```
I also suggest `self._available = True` in `XiaomiCoordinatedMiioEntity.__init__()`.
```suggestion
        self._available = True
        self._current_option = self._extract_value_from_attribute(
            self.coordinator.data, self._controller.short_name
        )
        self.async_write_ha_state()
```

jbouwh committed:
Update homeassistant/components/xiaomi_miio/select.py

Remove default value

Co-authored-by: Maciej Bieniek <bieniu@users.noreply.github.com>

jbouwh committed:
Update homeassistant/components/xiaomi_miio/select.py

Remove unsused self._state assignment

Co-authored-by: Maciej Bieniek <bieniu@users.noreply.github.com>

jbouwh committed:
Update homeassistant/components/xiaomi_miio/select.py

Remove checks for coordinator singe we are only available with a successful data fetch.

Co-authored-by: Maciej Bieniek <bieniu@users.noreply.github.com>

jbouwh committed:
Update homeassistant/components/xiaomi_miio/select.py

No need to check for fr coordinator since this call is triggerd at a succesfull update

Co-authored-by: Maciej Bieniek <bieniu@users.noreply.github.com>

jbouwh committed:
Merge remote-tracking branch 'upstream/rework-xiaomi-miio-fans-humidifiers' into add-xiaomi-miio-humidifier-select

jbouwh committed:
Do not override self.available use datacoordinator

bieniu reviewed (approved):
LGTM 👍

jbouwh committed:
Merge remote-tracking branch 'upstream/rework-xiaomi-miio-fans-humidifiers' into add-xiaomi-miio-humidifier-select

jbouwh committed:
Clean-up code

jbouwh committed:
Remove double feature check - clean up

jbouwh committed:
Remove unneeded attribute

jbouwh committed:
sort HUMIDIFIER_PLATFORMS

MartinHjelmare commented on homeassistant/components/xiaomi_miio/select.py:
```diff
@@ -0,0 +1,208 @@
+
+ATTR_LED_BRIGHTNESS = "led_brightness"
+
+SERVICE_TO_METHOD = {
```
I think we can wait with adding this abstraction until we have a second use case.

MartinHjelmare commented on homeassistant/components/xiaomi_miio/select.py:
```diff
@@ -0,0 +1,208 @@
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up the Selectors from a config entry."""
+    entities = []
+    if config_entry.data[CONF_FLOW_TYPE] == CONF_DEVICE:
```
Invert and return if true, etc.

jbouwh replied:
Done

MartinHjelmare commented on homeassistant/components/xiaomi_miio/select.py:
```diff
@@ -0,0 +1,208 @@
+        device = hass.data[DOMAIN][config_entry.entry_id][KEY_DEVICE]
+        coordinator = hass.data[DOMAIN][config_entry.entry_id][KEY_COORDINATOR]
+        device_features = 0
+        entity_class = None
```
```suggestion
```

MartinHjelmare commented on homeassistant/components/xiaomi_miio/select.py:
```diff
@@ -0,0 +1,208 @@
+
+        for feature in SELECTOR_TYPES:
+            selector = SELECTOR_TYPES[feature]
+            if feature & device_features and feature in SELECTOR_TYPES:
```
Do we even need this check? Don't we already know that the models above support an entity?

jbouwh replied:
Removed check and evaluation of device_features for now

MartinHjelmare commented on homeassistant/components/xiaomi_miio/select.py:
```diff
@@ -0,0 +1,208 @@
+    @property
+    def current_option(self):
+        """Return the current option."""
+        return getattr(self, SERVICE_TO_METHOD[self._controller.service]["property"])
```
Simplify this for now and remove all of the abstraction.

jbouwh replied:
Done

MartinHjelmare commented on homeassistant/components/xiaomi_miio/select.py:
```diff
@@ -0,0 +1,208 @@
+        model = config_entry.data[CONF_MODEL]
+        device = hass.data[DOMAIN][config_entry.entry_id][KEY_DEVICE]
+        coordinator = hass.data[DOMAIN][config_entry.entry_id][KEY_COORDINATOR]
+        device_features = 0
```
```suggestion
```

jbouwh committed:
Update homeassistant/components/xiaomi_miio/select.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

jbouwh committed:
Update homeassistant/components/xiaomi_miio/select.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

jbouwh committed:
Remove service abstraction

jbouwh committed:
Remove extra space

jbouwh committed:
Retrigger tests

MartinHjelmare commented on homeassistant/components/xiaomi_miio/select.py:
```diff
@@ -0,0 +1,187 @@
+    @property
+    def led_brightness(self):
+        """Return the current led brightness."""
+        reversed_value_map = {0: "Bright", 1: "Dim", 2: "Off"}
```
This dict can be extracted to a constant.

jbouwh replied:
Added a const

MartinHjelmare commented on homeassistant/components/xiaomi_miio/select.py:
```diff
@@ -0,0 +1,187 @@
+
+    async def async_set_led_brightness(self, brightness: str):
+        """Set the led brightness."""
+        value_map = {"Bright": 0, "Dim": 1, "Off": 2}
```
This too.

jbouwh replied:
Added a const

MartinHjelmare commented on homeassistant/components/xiaomi_miio/select.py:
```diff
@@ -0,0 +1,187 @@
+    def led_brightness(self):
+        """Return the current led brightness."""
+        reversed_value_map = {0: "Bright", 1: "Dim", 2: "Off"}
+        if self._current_led_brightness in reversed_value_map:
```
We can just use `dict.get` instead of this check.

jbouwh replied:
Done

jbouwh committed:
Dict to const and use ditc.get()

jbouwh committed:
Merge branch 'rework-xiaomi-miio-fans-humidifiers' into add-xiaomi-miio-humidifier-select

MartinHjelmare commented on homeassistant/components/xiaomi_miio/select.py:
```diff
@@ -0,0 +1,183 @@
+    @property
+    def led_brightness(self):
+        """Return the current led brightness."""
+        return dict.get(LED_BRIGHTNESS_REVERSE_MAP, self._current_led_brightness)
```
```suggestion
        return LED_BRIGHTNESS_REVERSE_MAP.get(self._current_led_brightness)
```

MartinHjelmare commented on homeassistant/components/xiaomi_miio/select.py:
```diff
@@ -0,0 +1,183 @@
+    @property
+    def led_brightness(self):
+        """Return the current led brightness."""
+        return dict.get(LED_BRIGHTNESS_REVERSE_MAP_MIOT, self._current_led_brightness)
```
```suggestion
        return LED_BRIGHTNESS_REVERSE_MAP_MIOT.get(self._current_led_brightness)
```

MartinHjelmare commented on homeassistant/components/xiaomi_miio/select.py:
```diff
@@ -0,0 +1,183 @@
+
+LED_BRIGHTNESS_MAP = {"Bright": 0, "Dim": 1, "Off": 2}
+LED_BRIGHTNESS_MAP_MIOT = {"Bright": 2, "Dim": 1, "Off": 0}
+LED_BRIGHTNESS_REVERSE_MAP = {0: "Bright", 1: "Dim", 2: "Off"}
```
```suggestion
LED_BRIGHTNESS_REVERSE_MAP = {val: key for key, val in LED_BRIGHTNESS_MAP.items()}
```

MartinHjelmare commented on homeassistant/components/xiaomi_miio/select.py:
```diff
@@ -0,0 +1,183 @@
+LED_BRIGHTNESS_MAP = {"Bright": 0, "Dim": 1, "Off": 2}
+LED_BRIGHTNESS_MAP_MIOT = {"Bright": 2, "Dim": 1, "Off": 0}
+LED_BRIGHTNESS_REVERSE_MAP = {0: "Bright", 1: "Dim", 2: "Off"}
+LED_BRIGHTNESS_REVERSE_MAP_MIOT = {0: "Off", 1: "Dim", 2: "Bright"}
```
```suggestion
LED_BRIGHTNESS_REVERSE_MAP_MIOT = {val: key for key, val in LED_BRIGHTNESS_MAP_MIOT.items()}
```

jbouwh committed:
Update homeassistant/components/xiaomi_miio/select.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

jbouwh committed:
Update homeassistant/components/xiaomi_miio/select.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

jbouwh committed:
Update homeassistant/components/xiaomi_miio/select.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

jbouwh committed:
Update homeassistant/components/xiaomi_miio/select.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

MartinHjelmare reviewed (approved):
Looks good!

MartinHjelmare commented on homeassistant/components/xiaomi_miio/select.py:
```diff
@@ -0,0 +1,183 @@
+LED_BRIGHTNESS_MAP = {"Bright": 0, "Dim": 1, "Off": 2}
+LED_BRIGHTNESS_MAP_MIOT = {"Bright": 2, "Dim": 1, "Off": 0}
+LED_BRIGHTNESS_REVERSE_MAP = {val: key for key, val in LED_BRIGHTNESS_MAP.items()}
+LED_BRIGHTNESS_REVERSE_MAP_MIOT = {val: key for key, val in LED_BRIGHTNESS_MAP_MIOT.items()}
```
```suggestion
LED_BRIGHTNESS_REVERSE_MAP_MIOT = {
    val: key for key, val in LED_BRIGHTNESS_MAP_MIOT.items()
}
```

MartinHjelmare committed:
Black
